kolibri/core/logger/utils/exam_log_migration.py

```diff
@@ -108,7 +117,126 @@ def _update_attempt_logs(masterylog_id, logs):
+    data = examattemptlog.serialize()
+    for f in exam_attempts_blocklist:
+        if f in data:
+            del data[f]
+    data["sessionlog_id"] = sessionlog_id
+    data["masterylog_id"] = masterylog_id
+    attemptlog = AttemptLog.deserialize(data)
+    attemptlog.item = "{}:{}".format(examattemptlog.content_id, attemptlog.item)
+    attemptlog.id = attemptlog.calculate_uuid()
+    return attemptlog
+
+
+def _handle_examlog(examlog, unprocessed_attempt_log_ids):
+    examattemptlogs = examlog.attemptlogs.all()
+    try:
+        start_timestamp = min(e.start_timestamp for e in examattemptlogs)
+    except ValueError:
+        start_timestamp = local_now()
+    content_id = examlog.exam_id
+    user = examlog.user
+    complete = examlog.closed
+    progress = 1 if examlog.closed else 0
+    completion_timestamp = examlog.completion_timestamp
+    end_timestamp = examlog.completion_timestamp
+    dataset_id = user.dataset_id
+    session_log = ContentSessionLog(
+        user=user,
+        content_id=content_id,
+        start_timestamp=start_timestamp,
+        end_timestamp=end_timestamp,
+        progress=progress,
+        kind=kind,
+        dataset_id=dataset_id,
+    )
+    session_log.id = session_log.calculate_uuid()
+
+    summary_log = ContentSummaryLog(
+        user=user,
+        content_id=content_id,
+        start_timestamp=start_timestamp,
+        end_timestamp=end_timestamp,
+        completion_timestamp=completion_timestamp,
+        progress=progress,
+        kind=kind,
+        dataset_id=dataset_id,
+    )
+    summary_log.id = summary_log.calculate_uuid()
+
+    mastery_log = MasteryLog(
+        user=user,
+        summarylog_id=summary_log.id,
+        mastery_criterion=mastery_criterion,
+        start_timestamp=start_timestamp,
+        end_timestamp=end_timestamp,
+        completion_timestamp=completion_timestamp,
+        # Do this rather than just generating a random value
+        # so that the mastery log id is deterministic.
+        mastery_level=int(str(int(content_id, 16))[-9:]),
+        complete=complete,
+        dataset_id=dataset_id,
+    )
+    mastery_log.id = mastery_log.calculate_uuid()
+
+    attempt_logs = []
+
+    for examattemptlog in examattemptlogs:
+        attemptlog = _create_attemptlog(examattemptlog, session_log.id, mastery_log.id)
+        attempt_logs.append(attemptlog)
+        try:
+            unprocessed_attempt_log_ids.remove(examattemptlog.id)
+        except KeyError:
+            pass
+    return session_log, summary_log, mastery_log, attempt_logs
+
+
+def _handle_unprocessed_attemptlog_ids(unprocessed_attempt_log_ids):
+    if unprocessed_attempt_log_ids:
+        examlog_id_to_masterylog = {}
+        unprocessed_attempt_log_ids = list(unprocessed_attempt_log_ids)
+        attempt_logs = []
+        i = 0
+        source_logs = ExamAttemptLog.objects.filter(
+            id__in=unprocessed_attempt_log_ids[i : i + BATCH_READ_SIZE]
```

Comment:
It looks like this is doing batching but I think it only does one batch?

Reply:
Yes, I apparently forgot to while here.

Review saves lives again.

Reply:
Updated to actually loop.